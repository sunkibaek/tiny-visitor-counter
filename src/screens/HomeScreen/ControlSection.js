import { View, StyleSheet } from "react-native";

import Button from "../../components/Button";

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    paddingHorizontal: 16,
  },
});

const ControlSection = () => (
  <View style={styles.container}>
    <Button.Link title="Exit" fullWidth={true} />

    <Button title="Enter" fullWidth={true} />
  </View>
);

export default ControlSection;
