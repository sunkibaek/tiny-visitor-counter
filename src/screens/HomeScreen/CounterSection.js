import { StyleSheet, Text, View } from "react-native";

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
  },
  currentVisitorText: {
    fontSize: 128,
    fontWeight: "bold",
  },
  divider: {
    fontSize: 64,
    lineHeight: 128,
  },
  maxVisitorText: {
    fontSize: 64,
    lineHeight: 128,
    marginBottom: -8,
  },
  row: {
    flexDirection: "row",
    alignItems: "flex-end",
  },
  circle: {
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#C4C4C4",
    width: 360,
    height: 360,
    borderRadius: 180,
  },
  warningText: {
    color: "#FFC107",
  },
  errorText: {
    color: "#F44336",
  },
});

const CounterSection = ({ current, max, currentStatus }) => {
  return (
    <View style={styles.container}>
      <View style={styles.circle}>
        <View style={styles.row}>
          <Text
            style={[
              styles.currentVisitorText,
              currentStatus.equalToMax && styles.warningText,
              currentStatus.overMax && styles.errorText,
            ]}
          >
            {current}
          </Text>

          <Text style={styles.divider}> / </Text>
          <Text style={styles.maxVisitorText}>{max}</Text>
        </View>
      </View>
    </View>
  );
};

export default CounterSection;
