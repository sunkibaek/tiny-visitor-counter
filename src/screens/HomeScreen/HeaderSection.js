import { Button, StyleSheet, View } from "react-native";

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "flex-end",
    paddingHorizontal: 8,
  },
});

const HeaderSection = () => (
  <View style={styles.container}>
    <Button title="Reset" />
  </View>
);

export default HeaderSection;
