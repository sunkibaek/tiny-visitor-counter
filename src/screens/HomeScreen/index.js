import { useState } from "react";
import { Alert, SafeAreaView, StyleSheet, View } from "react-native";

import HeaderSection from "./HeaderSection";
import CounterSection from "./CounterSection";
import ControlSection from "./ControlSection";
import { DEFAULTS, getCurrentStatus } from "../../models/Counter";

const styles = StyleSheet.create({
  safeArea: {
    paddingVertical: 16,
    flex: 1,
  },
  container: {
    flex: 1,
  },
});

const HomeScreen = () => {
  const [max, setMax] = useState(DEFAULTS.max);
  const [current, setCurrent] = useState(0);

  const increaseCurrent = () => {
    setCurrent((prev) => {
      const newCurrent = prev + 1;

      if (newCurrent > DEFAULTS.limit) {
        return prev;
      }

      return newCurrent;
    });
  };

  const decreaseCurrent = () => {
    setCurrent((prev) => {
      if (prev === 0) {
        return 0;
      }

      return prev - 1;
    });
  };

  const resetAll = () => {
    Alert.alert("Reset", "Are you sure you want to reset?", [
      {
        text: "Cancel",
        style: "cancel",
      },
      {
        text: "Reset",
        onPress: () => {
          Alert.prompt("Reset", "Enter new max", (text) => {
            const newMax = parseInt(text, 10);

            if (Number.isNaN(newMax)) {
              Alert.alert("Invalid maxâ€”please enter a number");
              return;
            }

            setMax(newMax);
            setCurrent(0);
          });
        },
      },
    ]);
  };

  return (
    <SafeAreaView style={styles.safeArea}>
      <View style={styles.container}>
        <HeaderSection onResetPress={resetAll} />

        <CounterSection
          current={current}
          max={max}
          currentStatus={getCurrentStatus(current, max)}
        />

        <ControlSection
          onEnterPress={increaseCurrent}
          onExitPress={decreaseCurrent}
        />
      </View>
    </SafeAreaView>
  );
};

export default HomeScreen;
