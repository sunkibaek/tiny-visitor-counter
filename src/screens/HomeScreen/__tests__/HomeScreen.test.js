import { render, screen, fireEvent } from "@testing-library/react-native";

import HomeScreen from "../";

let mockDefaults;

jest.mock("../../../models/Counter", () => {
  mockDefaults = {
    max: 3,
    limit: 4,
  };

  return {
    ...jest.requireActual("../../../models/Counter"),
    DEFAULTS: mockDefaults,
  };
});

describe("HomeScreen", () => {
  it("renders correctly", () => {
    render(<HomeScreen />);

    expect(screen).toMatchSnapshot();
  });

  describe("counter", () => {
    it("increments the counter when the Enter button is pressed", () => {
      render(<HomeScreen />);
      const enterButton = screen.getByText("Enter");

      fireEvent.press(enterButton);

      expect(screen.getByTestId("currentVisitorText")).toBeTruthy();
    });

    it("decrements the counter when the Exit button is pressed", () => {
      render(<HomeScreen />);
      const enterButton = screen.getByText("Enter");
      const exitButton = screen.getByText("Exit");

      fireEvent.press(enterButton);
      fireEvent.press(exitButton);

      expect(screen.getByTestId("currentVisitorText")).toBeTruthy();
    });

    it("does not decrement the counter below 0", () => {
      render(<HomeScreen />);
      const exitButton = screen.getByText("Exit");

      fireEvent.press(exitButton);

      expect(screen.getByTestId("currentVisitorText")).toBeTruthy();
    });

    it("does not increment the counter above the limit", () => {
      mockDefaults.limit = 4;
      mockDefaults.max = 3;
      render(<HomeScreen />);
      const enterButton = screen.getByText("Enter");

      for (let i = 0; i < mockDefaults.limit + 1; i += 1) {
        fireEvent.press(enterButton);
      }

      expect(screen.getByTestId("currentVisitorText").props.children).toBe(
        mockDefaults.limit
      );
    });

    it("applies the warning color when the current counter equals to the max counter", () => {
      render(<HomeScreen />);
      const enterButton = screen.getByText("Enter");

      for (let i = 0; i < mockDefaults.max; i += 1) {
        fireEvent.press(enterButton);
      }

      expect(screen).toMatchSnapshot();
    });

    it("applies the danger color when the current counter is greater than the max counter", () => {
      render(<HomeScreen />);
      const enterButton = screen.getByText("Enter");

      for (let i = 0; i < mockDefaults.max + 1; i += 1) {
        fireEvent.press(enterButton);
      }

      expect(screen).toMatchSnapshot();
    });
  });
});
