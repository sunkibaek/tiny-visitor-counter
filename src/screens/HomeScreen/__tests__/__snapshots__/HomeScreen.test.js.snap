// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen counter applies the danger color when the current counter is greater than the max counter 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            undefined,
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Reset
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#C4C4C4",
            "borderRadius": 180,
            "height": 360,
            "justifyContent": "center",
            "width": 360,
          }
        }
      >
        <Text
          includeFontPadding={false}
          style={
            Array [
              Object {
                "fontSize": 128,
                "fontVariant": Array [
                  "tabular-nums",
                ],
                "fontWeight": "bold",
              },
              Object {
                "color": "#F44336",
              },
            ]
          }
          testID="currentVisitorText"
        >
          4
        </Text>
        <Text
          style={
            Object {
              "backgroundColor": "#000000",
              "height": 1,
              "width": 160,
            }
          }
        />
        <Text
          includeFontPadding={false}
          style={
            Object {
              "fontSize": 64,
              "fontVariant": Array [
                "tabular-nums",
              ],
              "marginTop": 16,
            }
          }
          testID="maxVisitorText"
        >
          3
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingHorizontal": 16,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Exit
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "#141414",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#ffffff",
              },
            ]
          }
        >
          Enter
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`HomeScreen counter applies the warning color when the current counter equals to the max counter 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            undefined,
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Reset
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#C4C4C4",
            "borderRadius": 180,
            "height": 360,
            "justifyContent": "center",
            "width": 360,
          }
        }
      >
        <Text
          includeFontPadding={false}
          style={
            Array [
              Object {
                "fontSize": 128,
                "fontVariant": Array [
                  "tabular-nums",
                ],
                "fontWeight": "bold",
              },
              Object {
                "color": "#FFC107",
              },
            ]
          }
          testID="currentVisitorText"
        >
          3
        </Text>
        <Text
          style={
            Object {
              "backgroundColor": "#000000",
              "height": 1,
              "width": 160,
            }
          }
        />
        <Text
          includeFontPadding={false}
          style={
            Object {
              "fontSize": 64,
              "fontVariant": Array [
                "tabular-nums",
              ],
              "marginTop": 16,
            }
          }
          testID="maxVisitorText"
        >
          3
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingHorizontal": 16,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Exit
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "#141414",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#ffffff",
              },
            ]
          }
        >
          Enter
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`HomeScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            undefined,
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Reset
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#C4C4C4",
            "borderRadius": 180,
            "height": 360,
            "justifyContent": "center",
            "width": 360,
          }
        }
      >
        <Text
          includeFontPadding={false}
          style={
            Array [
              Object {
                "fontSize": 128,
                "fontVariant": Array [
                  "tabular-nums",
                ],
                "fontWeight": "bold",
              },
              Object {},
            ]
          }
          testID="currentVisitorText"
        >
          0
        </Text>
        <Text
          style={
            Object {
              "backgroundColor": "#000000",
              "height": 1,
              "width": 160,
            }
          }
        />
        <Text
          includeFontPadding={false}
          style={
            Object {
              "fontSize": 64,
              "fontVariant": Array [
                "tabular-nums",
              ],
              "marginTop": 16,
            }
          }
          testID="maxVisitorText"
        >
          3
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingHorizontal": 16,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#141414",
              },
            ]
          }
        >
          Exit
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "#141414",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
              Object {
                "color": "#ffffff",
              },
            ]
          }
        >
          Enter
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
